Send wxEVT_SLIDER event only when slider value actually changes

PBfordev opened the pull request:
Due to wxMSW implementation detail, two wxEVT_SLIDER events were always
generated for one change in the slider value. Do not do this anymore,
generate wxEVT_SLIDER only when the slider value actually changes.

See [ticket 18929](https://trac.wxwidgets.org/ticket/18929).

**EDIT**
The first "fix" was wrong, checking for `wxEVT_SCROLL_CHANGED` instead of `SB_ENDSCROLL`. It hopefully works as intended now; however, as always I failed to squash the commits.

PBfordev committed:
Send wxEVT_SLIDER event only when slider value actually changes

Due to wxMSW implementation detail, two wxEVT_SLIDER events were always
generated for one change in the slider value. Do not do this anymore,
generate wxEVT_SLIDER only when the slider value actually changes.

PBfordev committed:
Send wxEVT_SLIDER event only when slider value actually changes

Due to wxMSW implementation detail, two wxEVT_SLIDER events were always
generated for one change in the slider value. Do not do this anymore,
generate wxEVT_SLIDER only when the slider value actually changes.

Closes #18929.

PBfordev committed:
Send wxEVT_SLIDER event only when slider value actually changes

Due to wxMSW implementation detail, two wxEVT_SLIDER events were always
generated for one change in the slider value. Do not do this anymore,
generate wxEVT_SLIDER only when the slider value actually changes.

Closes #18929.

vadz commented on src/msw/slider.cpp:
```diff
@@ -313,15 +313,28 @@ bool wxSlider::MSWOnScroll(int WXUNUSED(orientation),
+    // parameter is SB_ENDSCROLL, which always follows only after
+    // another scroll message which already changed the slider value.
+    // Therefore, sending wxEVT_SLIDER after SB_ENDSCROLL
+    // would result into two wxEVT_SLIDER events with the same value.
```
```suggestion
    // would result in two wxEVT_SLIDER events with the same value.
```

vadz commented on src/msw/slider.cpp:
```diff
@@ -313,15 +313,28 @@ bool wxSlider::MSWOnScroll(int WXUNUSED(orientation),
     SetValue(newPos);
 
     wxScrollEvent event(scrollEvent, m_windowId);
+    bool          result = false;
```
I'd rename it to
```suggestion
    bool          processed = false;
```
because this seems to better describe what it corresponds to.

vadz reviewed (approved):
Thanks! This doesn't do what I suggested (I wanted to do what you wrote, i.e. avoid duplicate events with the same position), but if this works, all the better. I think I was confused because I thought we were getting `SB_THUMBPOSITION`, but `SB_ENDSCROLL` makes much more sense, so I must have been just mistaken.

Anyhow, if this fixes the problem you've described, it should certainly be applied, thanks again!

The only thing is that, as usual, ideal would be to have a test for it, but I guess implementing the scenario you described using `wxUIActionSimulator` is rather non-trivial...

PBfordev commented:
Thanks for the quick reply.

>I wanted to do what you wrote, i.e. avoid duplicate events with the same position

I wanted to do that but I did not know how (short of introducing a member variable), as when we handle the message, the control already has the new value and there is nothing to compare it with?

Regarding the test. I may be missing something but looking at the current `wxSlider` tests, to test if we get a single `wxEVT_SLIDER` event instead of two after e.g. pressing `<Right>` seems to be simple but I am not sure it would be useful.

BTW, the docs for `wxEVT_SCROLL_CHANGED` seem to be somewhat incorrect, at least for `wxSlider` (`wxScrollBar` is not used in widgets sample so I did not try it for now). This event seems to be the only event generated when the value is changed when rolling the mouse wheel over the slider, see [the code](https://github.com/wxWidgets/wxWidgets/blob/master/src/msw/slider.cpp#L279) for handling `SB_THUMBPOSITION` in `wxSlider::MSWOnScroll()`.

PBfordev committed:
Update src/msw/slider.cpp

Co-authored-by: VZ <vz-github@zeitlins.org>

PBfordev committed:
Send wxEVT_SLIDER event only when slider value actually changes

Due to wxMSW implementation detail, two wxEVT_SLIDER events were always
generated for one change in the slider value. Do not do this anymore,
generate wxEVT_SLIDER only when the slider value actually changes.

Closes #18929.

vadz commented:
> > I wanted to do what you wrote, i.e. avoid duplicate events with the same position
> 
> I wanted to do that but I did not know how (short of introducing a member variable), as when we handle the message, the control already has the new value and there is nothing to compare it with?

It doesn't really matter, your fix is better anyhow as it's simpler and more clear.

> Regarding the test. I may be missing something but looking at the current `wxSlider` tests, to test if we get a single `wxEVT_SLIDER` event instead of two after e.g. pressing `<Right>` seems to be simple but I am not sure it would be useful.

The problem is that we need to press the mouse on the slider thumb to test the exact situation you're describing, aren't you? Or have I been misunderstanding you since the beginning and did you mean that you could release the mouse after clicking when you wrote "Click on the slider thumb and then repeatedly press right arrow"? I've been testing with pressing the arrow _while_ keeping the mouse button pressed... Retrying it right now I indeed see that the same problem happens if I just press the arrow without doing anything with the mouse, sorry for complete reading comprehension failure :-(

Anyhow, yes, it would be simple to add a unit test for this: just give the focus to the control and emulate two key presses and use `EventCounter` to check that we get the expected number of events. It would be great if you could do it, TIA!

> BTW, the docs for `wxEVT_SCROLL_CHANGED` seem to be somewhat incorrect, at least for `wxSlider` (`wxScrollBar` is not used in widgets sample so I did not try it for now). This event seems to be the only event generated when the value is changed when rolling the mouse wheel over the slider, see [the code](https://github.com/wxWidgets/wxWidgets/blob/master/src/msw/slider.cpp#L279) for handling `SB_THUMBPOSITION` in `wxSlider::MSWOnScroll()`. Perhaps the docs should be updated to reflect this (I would do that)?

I'm not sure if it's the documentation or the code which is wrong. The behaviour described in the docs makes sense to me, i.e. you should be able to catch `wxEVT_SCROLL_CHANGED` to get the notification about the final position change independently of the way it happens. If it doesn't get sent, we probably should send it.

> **Edit:** ` wxScrollBar` does not respond to the mouse wheel.

I'm rather surprised that it doesn't, this could be another bug... But let's leave it aside for now.

PBfordev committed:
Add test case for wxEVT_SLIDER

PBfordev commented:
> It doesn't really matter, your fix is better anyhow as it's simpler and more clear.

JFYI, there can still be multiple `wxEVT_SLIDER` and same `wxEVT_SCROLL*` events with the same value. But it happens only as the result of user action, when they keep ramming against the set minimum or maximum value. The same happens for `wxScrollBar` so I think this is expected, or at least acceptable, behaviour.

>I've been testing with pressing the arrow while keeping the mouse button pressed

Sorry for being unclear, the mouse click was there just to give the slider focus so it can be controlled from keyboard.

> Anyhow, yes, it would be simple to add a unit test for this: just give the focus to the control and emulate two key presses and use `EventCounter` to check that we get the expected number of events. 

Done.

> I'm not sure if it's the documentation or the code which is wrong. The behaviour described in the docs makes sense to me, i.e. you should be able to catch `wxEVT_SCROLL_CHANGED` to get the notification about the final position change independently of the way it happens. If it doesn't get sent, we probably should send it.

Once again, sorry for being unclear. There are several scroll events (depending on user actions), changing the slider value. These are all always followed by `wxEVT_SCROLL_CHANGED` with the same value, which is generated because of handling `SB_ENDSCROLL` (and this was the cause of duplicate  `wxEVT_SLIDER`). However, when the mouse wheel is used to change the slider value, unlike in any other situation `wxEVT_SCROLL_CHANGED`  is not preceded by another scroll event, see the link in my previous post why is that. That is why I found its description saying "end of scrolling events" odd, as there is a case where it does not end scrolling events.

BTW, I wonder whether the slider page in widgets sample should not also show `wxEVT_SLIDER` , to me it would make sense if it did, i.e., do what my patched sample does. On a similar note, perhaps it could be a good idea to use a splitter between the log window and the rest of the sample frame: As it is now, I can never see more than 3 lines of log at once, no matter how big the frame is.

vadz commented:
> JFYI, there can still be multiple `wxEVT_SLIDER` and same `wxEVT_SCROLL*` events with the same value. But it happens only as the result of user action, when they keep ramming against the set minimum or maximum value. The same happens for `wxScrollBar` so I think this is expected, or at least acceptable, behaviour.

Yes, I agree.

> Once again, sorry for being unclear. There are several scroll events (depending on user actions), changing the slider value. These are all always followed by `wxEVT_SCROLL_CHANGED` with the same value, which is generated because of handling `SB_ENDSCROLL` (and this was the cause of duplicate `wxEVT_SLIDER`). However, when the mouse wheel is used to change the slider value, unlike in any other situation `wxEVT_SCROLL_CHANGED` is not preceded by another scroll event, see the link in my previous post why is that. That is why I found its description saying "end of scrolling events" odd, as there is a case where it does not end scrolling events since there were none before.

Ah, I see it now, thanks for taking time to explain. I guess we could indeed be more explicit about this case, although, to be honest, I don't see how it could create any problems in practice (i.e. why would you rely on getting some other event before `wxEVT_SCROLL_CHANGED` -- the important thing, IMO, is that you're guaranteed to get this one at the end).

> BTW, I wonder whether the slider page in widgets sample should not also show `wxEVT_SLIDER` , to me it would make sense if it did, i.e., do what my patched sample on wxTrac does.

Yes, agreed, so I'll just apply it as part of merging this PR (I have to squash it anyhow), thanks!

> On a similar note, perhaps it could be a good idea to use a splitter between the log window and the rest of the sample frame: As it is now, I can never see more than 3 lines of log at once, no matter how big the frame is.

I agree that this is something that could/should be improved too, but I'm not sure if splitter will help that much, as there is just not enough vertical space to show everything on all pages. Maybe we should just have a separate log window? In any case, this is not really related of this PR, so I'll merge this one right now, but any future PRs changing this would be welcome, of course.

Thanks again!

PBfordev commented:
> > BTW, I wonder whether the slider page in widgets sample should not also show `wxEVT_SLIDER` , to me it would make sense if it did, i.e., do what my patched sample on wxTrac does.
> 
> Yes, agreed, so I'll just apply it as part of merging this PR (I have to squash it anyhow), thanks!


Sorry, I did not mean to take the sample patch as it was, I meant to take it in spirit: I promise to try harder to make myself clear. The wxTrac sample patch has a "bug", where the two `OnSlider()` event handlers each keep their own slider event count where IMO there should be just one. I attempted to fix this in #2082.

Regarding the widgets sample sizing, TBH, IMO the page with a widget should not be a` wxPanel` but `wxScrolledWindow`, to ensure none of the GUI is ever inaccessible when the window is too small.
